proposals/conditional_expressions/cond.txt

```diff
@@ -0,0 +1,34 @@
+To: J3                                                     J3/XXX
+From: Ondrej Certik
+Subject: Conditional expressions using intrinsic function
+References: 21-157, 21-159
+Date: 2021-June-26
+
+Paper 21-157 proposes two syntactic forms for
+conditional expressions. The paper 21-159 proposes
+an alternative form. This paper proposes to implement
+conditional expressions using intrinsic functions
+
+The syntactics forms in 21-157 is too verbose, as argued in 21-159. The form in
+21-159 is concise, but based on the poll [1], it is not the most popular
+either.
+
+We propose the following syntactic form
+
+   <cond-expr> is ifthen( predicate, consequent,
+                      [ ifthen( predicate, consequent, ]*
+                      alternative) [)]*
```

Comment:
I don't think this syntax is correct. In this syntax, only the first two arguments are required. But what is the result of:

```
res = ifthen(.false., 2)
```

I think the correct syntax is:

```suggestion
   <cond-expr> is ifthen( predicate, consequent, alternative )
```

and then one can use another call to `ifthen()` as the `alternative` if chaining is needed. Unless I'm missing some technical or semantic subtlety.

Reply:
I didn't know how to write it correctly, but the way it is written, `alternative` is out of `[]`, so it is not optional. So `ifthen(.false., 2)` is not allowed. The issue with `ifthen( predicate, consequent, alternative )` is that it might not be completely clear that in `ifthen( predicate, consequent, ifthen(predicate2, consequent2, alternative) )` the second `ifthen` is only evaluated if `predicate` is false. But I guess that follows from the general definition of `ifthen` that arguments are not evaluated unless needed and it applies recursively.

Reply:
You're right, somehow I didn't see `alternative` in your original form as I was writing the comment and example. What threw me off is that the optional arg comes before the required, which is not allowed. So if you put the `alternative` as the third argument, you don't need the thing in brackets.